internal/primitives/state-machine/overlayed_entry.go

```diff
@@ -0,0 +1,338 @@
+		oe.TransactionExtrinsics().insert(*atExtrinsic)
+	}
+}
+
+// Writes a new version of a value.
+// This makes sure that the old version is not overwritten and can be properly
+// rolled back when required.
+func (oe *OverlayedEntry[V]) Set(value StorageValue, firstWriteInTx bool, atExtrinsic *uint32) {
+	var action storageEntry
+
+	if value == nil {
+		action = removeStorageEntry{}
+	} else {
+		action = setStorageEntry{value}
+	}
+
+	if firstWriteInTx || len(oe.transactions) == 0 {
+		oe.transactions = append(oe.transactions, transaction[V]{
+			value:      action.(V),
+			extrinsics: extrinsics{},
+		})
+	} else {
+		oldValue := oe.ValueRef()
+
+		var setPrev *struct {
+			data                []byte
+			currentMaterialized *uint
+			parentSize          uint
+		}
+
+		switch oldVal := any(oldValue).(type) {
+		case appendStorageEntry:
+			*oldValue = any(oldVal).(V)
+
+			setPrev = &struct {
+				data                []byte
+				currentMaterialized *uint
+				parentSize          uint
+			}{
+				data:                []byte{},
+				currentMaterialized: oldVal.materializedLength,
+				parentSize:          *oldVal.parentSize,
+			}
+		}
+
+		*oldValue = action.(V)
+		if setPrev != nil {
+			transactions := len(oe.transactions)
+			if transactions < 2 {
+				panic("`set_prev` is not nil if the value came from parent; qed")
+			}
+
+			parent := &oe.transactions[transactions-2]
+
+			parentValue, ok := any(parent.value).(storageEntry)
+			if !ok {
+				panic("Parent value is not a storage entry")
+			}
+
+			restoreAppendToParent(parentValue, setPrev.data, setPrev.currentMaterialized, setPrev.parentSize)
+		}
+	}
+
+	if atExtrinsic != nil {
+		oe.TransactionExtrinsics().insert(*atExtrinsic)
+	}
+}
+
+// Append content to a value, updating a prefixed compact encoded length.
+// This makes sure that the old version is not overwritten and can be properly
+// rolled back when required.
+// This avoid copying value from previous transaction.
+func (oe *OverlayedEntry[V]) Append(
+	element StorageValue,
+	firstWriteInTx bool,
+	init func() StorageValue,
+	atExtrinsic *uint32,
+) {
+	var data []byte
+	var currentLength uint
+	var materializedLength *uint
+	var parentSize *uint
+
+	replace := true
+
+	if len(oe.transactions) == 0 {
+		initValue := init()
+		storageAppend := newStorageAppend(&initValue)
+
+		// Either the init value is a SCALE list like value to that the `element` gets appended
+		// or the value is reset to `element`.
+		length := storageAppend.extractLength()
+		if length != nil {
+			storageAppend.appendRaw(element)
+			data = initValue
+			currentLength = *length + 1
+			materializedLength = length
+		} else {
+			data = element
+			currentLength = 1
+			materializedLength = nil
+		}
+
+		oe.transactions = append(oe.transactions, transaction[V]{
+			value: any(&appendStorageEntry{
+				data:               data,
+				currentLength:      currentLength,
+				materializedLength: materializedLength,
+				parentSize:         nil,
+			}).(V),
```

Comment:
Since these methods are supposed to be only when `V` is of type `StorageEntry` we could encounter some side effects when using `OverlayedEntry`  and calling  `Set` or `Append` or `Value` in other use cases.  To essentially support `impl OverlayedEntry<StorageEntry>`  I was thinking you could wrap the original `OverlayedEntry` type and move these methods to that type were `V` is `StorageEntry`. You do something similar with `OverlayedChangeSet`

Reply:
I like this a lot better.